### DISCUSSION TITLE
Digit confusion between 1 and 7

### DISCUSSION THREAD

**Discussion by Mika-ncy** [jomjol/AI-on-the-edge-device#2637] [General]
### The Problem

Hi ! 

There is an issue on my working system. The digit 1 on my counter is missread into a 7.

Si there any solution to learn the system not to confuse ?

### Version

v15.3.0

### Logfile

```shell
[2d21h12m47s] 2023-09-28T09:48:52 <ERR> [POSTPROC] main: Raw: 074748.0, Value: , Status: Rate too high - Read: 74748.0 - Pre: 14148.0 - Rate: 60600.0
[2d21h15m46s] 2023-09-28T09:51:52 <ERR> [POSTPROC] main: Raw: 074748.0, Value: , Status: Rate too high - Read: 74748.0 - Pre: 14148.0 - Rate: 60600.0
[2d21h18m46s] 2023-09-28T09:54:52 <ERR> [POSTPROC] main: Raw: 074748.0, Value: , Status: Rate too high - Read: 74748.0 - Pre: 14148.0 - Rate: 60600.0
[2d21h21m49s] 2023-09-28T09:57:55 <ERR> [POSTPROC] main: Raw: N44NNN.5, Value: , Status: Rate too high - Read: 44148.5 - Pre: 14148.0 - Rate: 30000.5
[2d21h24m44s] 2023-09-28T10:00:50 <ERR> [POSTPROC] main: Raw: 074748.0, Value: , Status: Rate too high - Read: 74748.0 - Pre: 14148.0 - Rate: 60600.0
```

**Comment by caco3** [comment]
Please provide a screenshot of your device

**Comment by TopGun2010** [comment]
Hey - I got the same problem over here. Quite annoying. And yes, I see what the problem is - the lighting produces a reflection on the upper edge of the counter wheel. But I am not able to get the light at any other angle or whatever, then it might occour at a differenz number wheel and the problem is more to the right/left. I think this could be also solved by a better recognition - as a 1 looks differenz from a 7, the problem sometimes also evolves with other numbers e.g. a 2 is also recognized as a 7.

Using 15.7.0

![screen02](https://github.com/user-attachments/assets/f700abe2-a741-4170-a731-e85b35844847)
![screen01](https://github.com/user-attachments/assets/bbda937f-dc6e-4278-8369-7ab0daf4f67a)

Any chance for a better recognition?

**Reply by caco3** [reply]
the right way to solve this is to ged rid of the reflections. If not possible with the internal LED and a diffusor, you can try with external lighting.

**Reply by TopGun2010** [reply]
yes, for sure, but this makes the whole solution, which you can use quite "out of the box" by just installing the esp and flashing it - a lot more fiddly (soldering an extra led etc.) and might be in whole more stable, when the recognition algorithm/ai is better and can solve this.
just a point to consider and optimize. I mean sometimes - very rarely it does not recognize a number right on a wheel without reflection. Guess there is room for improvement.

**Reply by TopGun2010** [reply]
I could save more images and provide training data.

**Reply by SybexX** [reply]
It makes no sense to train new faulty images and this only increases the error rate even further.
The ESP has limited computing capacity and therefore only a limited range of functions from Tensorflow.
- Use version 16.0.0 and the zoom function implemented in it to increase the ROI size slightly.
- Increase CamGainceiling and reduce the LEDIntensity at the same time, this should help you eliminate most reflections
- and maybe play around with the other camera settings
- position the ROIs a little better

I'm starting to ask myself why the tooltips/wiki are there and why they're regularly maintained if no one pays attention to them anyway.

**Reply by TopGun2010** [reply]
Tried everything before and again now with RC5 of 16.0.0, seems as if I really have to get rid of the reflections anyhow.

**Reply by caco3** [reply]
If the image does not contain the needed information, the ESP can not invent it. It is like on a normal camera, if you have noise, there is no way to remove it by software (without degrading the other image quality).
The right step really is to get rid of the reflections.

**Reply by TopGun2010** [reply]
I understand what you're saying, but I was curious why the 1 and 7, which do look different are not recognized as such. I mean, you can see the small thing of the 1 on the top, the 7 has an angeled stroke, the 1 is straight. Are you trying to cover different font styles? Are there styles where the 7 also got a straigt stroke like the 1? This is a thing which I cannot understand, cause when this is given the reflection wouldn't matter. But dont wanna have a discussion here at this point - I gotta get rid of the reflections somehow.

**Reply by caco3** [reply]
Keep in mind that there is only one model for all trained digits.
The neuronal network just outputs the best fit. This does how  ever not mean that it is the right/expected result.

The model is a tradeoff between being generic (so it works for all kind of digits) and a good enough recognition.
The ideal ways would be to have a well trained model containing just those digits which you have. But this would be a lot of extra work.

**Reply by TopGun2010** [reply]
Okay and the tradeoff is as computing power is so limited I guess? And also of less training data? I'm not familiar with file sizes or how much more training data causes how much more computing power or much lager file sizes. 
I just saw what was possible when some ppl of our Astronomy discord programmed a neuronal network to determine and eliminate color/light gradients in astro pictures. Which was quite impressive and became known in the astro community world wide in just a few months, but - admittedly it uses PC/MAC CPU power or even GPU/cuda :D But its more complex and pictures are more likely in a 26 MP range. So I was falsely thinking there would be more possible here. :)

**Comment by friedpa** [comment]
@SybexX  The problem with the tool-tips/wiki is, that you and @caco3 spending a lot of time for explaining things over and over again. It seems, that a lot of users are simply lazy, why reading the tool-tips/ wiki, if I can get an answer without using any brain-power...
In other Git-hub forums there is a standard text send out pointing to the tool-tips/ wiki... just my fifty cents....

**Reply by caco3** [reply]
@friedpa Thanks for your feedback. Let us discuss this in a separate issue: https://github.com/jomjol/AI-on-the-edge-device/issues/3444

I would be happy for your feedback and if possible also contribution.

**Comment by TopGun2010** [comment]
Just to defend myself or proove that I also read it, I think the problem why many ppl are writing here might be something different. Might be my understanding. For me it is not logical, why zooming in the problem would lead to another readout. Also higher gain and less LED power results in the same result in my understanding. Coming from photography I know when I got less light, but more sensor sensitivity/amplification (gain), the reflection stays the same relatively to the brightness of the picture, only the noise raises significantly, results in less quality signal and recognition.

For me it is not a faulty image you would be training with, it is something where it should be recognized as well, just other circumstances. And I bet many ppl got very similar problems which would be solved then. But your argument concerning limited computing power is more than an argument for me. 

Don't get it wrong, love what you're doing, I just wanted to maybe get you into an "Apple-like" approach, to make it easy to use and kind of more fool proof. 

Thanx for your time!

**Reply by SybexX** [reply]
@TopGun2010 What many users forget is that it is not the human eye that evaluates the values ​​for the ESP, but the ESP itself.
What means poor image quality to your eye may mean very good quality to the ESP and vice versa.

**Reply by TopGun2010** [reply]
At least a more stable signal than for our eyes.. yeah. Thanx anyway. Just FYI the pic as it looks now, but the reflection disturbs it too much.

![zaehler](https://github.com/user-attachments/assets/49ad69a6-df14-460d-9bcb-486d1eabcdab)

**Reply by caco3** [reply]
This is odd, because I have a similar meter and have no issues although have also some reflections:
![grafik](https://github.com/user-attachments/assets/89f371e8-53bd-41b7-a524-f233052750a0)

**Reply by TopGun2010** [reply]
yeah but your reflection is smoother and at the bottom. guess down there are less problems than on top, where it could interfere with 1, 2 and 7 which are at least my problem digits. Guess the contrast and sharp edgedness of my reflection lets it look like a horizontal stroke at the top.

**Reply by SybexX** [reply]
This is what my picture looks like, just a catastrophe to the eye but exactly right for the ESP.
![Unbenannt](https://github.com/user-attachments/assets/c9a0377d-1681-419b-8e3a-8e8c6f1de7e0)

**Reply by caco3** [reply]
> yeah but your reflection is smoother

That's because I added a Diffusor to reduce the reflections.

**Reply by TopGun2010** [reply]
yeah I tried that, too.. but cannot get it that smooth as it is in your pic. It still remains some kind of strong stroke. Will try more and harder.. Just so fiddly.

thanx guys!

**Reply by caco3** [reply]
Do you have access to a 3D-Printer?
I am designing a case for the ESP32-CAM which will contain a diffusor.

**Reply by TopGun2010** [reply]
Unfortunately not. I will try again by diffusing it, maybe I can do better this time.

**Reply by TopGun2010** [reply]
Happy New Year btw :)

**Comment by Ingo-Knito** [comment]
I have one :-) Maybe I can help?

To eliminate reflections - in my optinion - it is best to use external LEDs.
I've designed a housing for the krom schröder BK G16 - which has the same housing as the G10.
Here with two external LED
![image](https://github.com/user-attachments/assets/042ebc39-ce60-434b-8f1e-9377ec81d979)

**Reply by caco3** [reply]
yes, that looks much better.
For my water and gas meters I work with the internal LED and the reflections are not disturbing (the gas meter has a diffusor made of white paper). But since so many people have issues I thought a 3D printed diffusor might be useful.

**Reply by TopGun2010** [reply]
Yeah, I know the best solution would be with external LED, but my initial point was, to make it as easy as it can be. With no soldering etc. But yeah, physics of reflections make it hard with the internal LED.

**Comment by TopGun2010** [comment]
A small update - luckily the affected 1 and 7 are before the decimal point and that far away from each other that I think, I was able to manage the faulty recognition by suppressing it with the error correction - when applying a much smaller change rate. Was working for at least one time now when crossing 0-2... lets see. Not as it's been ment to be, but seems to be working. 

"Softboxing" the internal LED didnt work for me, would have to place the whole esp32 elsewhere.
